### ISSUE TITLE
OpenIdConnectConfiguration not serializing properly

### ISSUE THREAD

**Issue by RogerioWagner** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#1475]
The OpenIdConnectConfiguration is not serializing the field **jwks_uri** field, the bellow example show the incorrect field naming (**jwksUri**).

The package to reproduce the behavior is Microsoft.IdentityModel.Protocols.OpenIdConnect version 5.5.0.

The problem caused is that the authenticationservice could not ask the authority the public key set to validate a token.

```json
{
  "additionalData": {},
  "acrValuesSupported": [],
  "authorizationEndpoint": "https://localhost:4000/connect/authorization",
  "checkSessionIframe": null,
  "claimsSupported": [], "claimsLocalesSupported": [],
  "claimsParameterSupported": false,
  "claimTypesSupported": [],
  "displayValuesSupported": [],
  "endSessionEndpoint": null,
  "frontchannelLogoutSessionSupported": null,
  "frontchannelLogoutSupported": null,
  "grantTypesSupported": [],
  "httpLogoutSupported": false,
  "idTokenEncryptionAlgValuesSupported": [],
  "idTokenEncryptionEncValuesSupported": [],
  "idTokenSigningAlgValuesSupported": [],
  "issuer": "https://localhost:4000",
  "jwksUri": "https://localhost:4000/.well-known/openid-configuration/jwks",
  "jsonWebKeySet": {
    "additionalData": {},
    "keys": [
      {
        "additionalData": {},
        "alg": null,
        "crv": null,
        "d": null,
        "dp": null,
        "dq": null,
        "e": "AQAB",
        "k": null,
        "keyId": "319D93355237260601A940A962991084B835681A",
        "keyOps": [],
        "kid": "319D93355237260601A940A962991084B835681A",
        "kty": "RSA",
        "n": "4wyItVfNjf/3NVwA8z1ijlAz/SmGSdQ+U2M22TwR/vQWGmT3Fivg7Do9otPwTQBTIVkU3NnTyhB8jH0pe2LSqkVJXzwQjtma3aH/GaHMYTArB36xf5YESPWb1tvVmbIFFxtqaMZYnvUOYqh4c6yYBLPbQkKE8aSvYtEfvucA5XPwQU11sA9fRuScWO0psKBD1xtXZDB3vzv0JSZFmWB6OSCQcrBiBLFsGUlDogOCuLqN8g5xxjUE7gv7hRnUu6sm9lhEdzTskDw5zjr0mh093/PZ1M0UYMCg/M6TZmz+mdrd3r0Yq97am6mdFACe2v2tNXryy2iLrVBred6W+g8tjQ==",
        "oth": null,
        "p": null,
        "q": null,
        "qi": null,
        "use": "sig",
        "x": null,
        "x5c": [],
        "x5t": "319D93355237260601A940A962991084B835681A",
        "x5tS256": null,
        "x5u": null,
        "y": null,
        "keySize": 2048,
        "hasPrivateKey": false,
        "cryptoProviderFactory": {
          "cryptoProviderCache": {},
          "customCryptoProvider": null,
          "cacheSignatureProviders": true
        }
      }
    ],
    "skipUnresolvedJsonWebKeys": true
  },
  "logoutSessionSupported": false,
  "opPolicyUri": null,
  "opTosUri": null,
  "registrationEndpoint": null,
  "requestObjectEncryptionAlgValuesSupported": [],
  "requestObjectEncryptionEncValuesSupported": [],
  "requestObjectSigningAlgValuesSupported": [],
  "requestParameterSupported": false,
  "requestUriParameterSupported": false,
  "requireRequestUriRegistration": false,
  "responseModesSupported": [],
  "responseTypesSupported": [],
  "serviceDocumentation": null,
  "scopesSupported": [],
  "signingKeys": [],
  "subjectTypesSupported": [],
  "tokenEndpoint": "https://localhost:4000/connect/authentication",
  "tokenEndpointAuthMethodsSupported": [],
  "tokenEndpointAuthSigningAlgValuesSupported": [],
  "uiLocalesSupported": [],
  "userInfoEndpoint": null,
  "userInfoEndpointEncryptionAlgValuesSupported": [],
  "userInfoEndpointEncryptionEncValuesSupported": [],
  "userInfoEndpointSigningAlgValuesSupported": []
}
```

**Comment by brentschmaltz** [comment]
@RogerioWagner the issue here is that the discovery spec doesn't include a jsonWebKeySet, i believe you will find the keys in additional data.

**Comment by RogerioWagner** [comment]
> @RogerioWagner the issue here is that the discovery spec doesn't include a jsonWebKeySet, i believe you will find the keys in additional data.

It's present in the json i've posted just because some experiments, currently it's not in the implementation anymore.

To workaround the problem I've made a copy of OpenIdConnectConfiguration.cs from the repository and changed the property JwksUri to jwks_uri.

        /// <summary>
        /// Gets or sets the 'jwks_uri'
        /// </summary>
        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore, NullValueHandling = NullValueHandling.Ignore, PropertyName = "jwks_uri", Required = Required.Default)]
        public string jwks_uri { get; set; }

The method implementation with OpenIdConnectConfiguration;

        [HttpGet]
        public IActionResult GetConfiguration()
        {
            
            // Ensure token and keys requests use the same cluster as this request
            string requestedUrlBase = $"{Request.Scheme}://{Request.Host}";

            //SeraphOpenIdConnectConfiguration configuration = new SeraphOpenIdConnectConfiguration();
            OpenIdConnectConfiguration configuration = new OpenIdConnectConfiguration();

            configuration.Issuer = requestedUrlBase;
            configuration.AuthorizationEndpoint = requestedUrlBase + "/" + SeraphConstants.AuthorizationRoute;
            configuration.TokenEndpoint = requestedUrlBase + "/" + SeraphConstants.TokenRoute;
            //configuration.jwks_uri = requestedUrlBase + Request.Path + "/" + SeraphConstants.PublicKeySetRoute;
            configuration.JwksUri = requestedUrlBase + Request.Path + "/" + SeraphConstants.PublicKeySetRoute; 

            return Ok(configuration);
        }

Will return: 

```json
{
  "additionalData": {},
  "acrValuesSupported": [],
  "authorizationEndpoint": "https://localhost:4000/connect/authorization",
  "checkSessionIframe": null,
  "claimsSupported": [],
  "claimsLocalesSupported": [],
  "claimsParameterSupported": false,
  "claimTypesSupported": [],
  "displayValuesSupported": [],
  "endSessionEndpoint": null,
  "frontchannelLogoutSessionSupported": null,
  "frontchannelLogoutSupported": null,
  "grantTypesSupported": [],
  "httpLogoutSupported": false,
  "idTokenEncryptionAlgValuesSupported": [],
  "idTokenEncryptionEncValuesSupported": [],
  "idTokenSigningAlgValuesSupported": [],
  "issuer": "https://localhost:4000",
  "jwksUri": "https://localhost:4000/.well-known/openid-configuration/jwks",
  "jsonWebKeySet": null,
  "logoutSessionSupported": false,
  "opPolicyUri": null,
  "opTosUri": null,
  "registrationEndpoint": null,
  "requestObjectEncryptionAlgValuesSupported": [],
  "requestObjectEncryptionEncValuesSupported": [],
  "requestObjectSigningAlgValuesSupported": [],
  "requestParameterSupported": false,
  "requestUriParameterSupported": false,
  "requireRequestUriRegistration": false,
  "responseModesSupported": [],
  "responseTypesSupported": [],
  "serviceDocumentation": null,
  "scopesSupported": [],
  "signingKeys": [],
  "subjectTypesSupported": [],
  "tokenEndpoint": "https://localhost:4000/connect/authentication",
  "tokenEndpointAuthMethodsSupported": [],
  "tokenEndpointAuthSigningAlgValuesSupported": [],
  "uiLocalesSupported": [],
  "userInfoEndpoint": null,
  "userInfoEndpointEncryptionAlgValuesSupported": [],
  "userInfoEndpointEncryptionEncValuesSupported": [],
  "userInfoEndpointSigningAlgValuesSupported": []
}
```

With SeraphOpenIdConnectConfiguration, the one with renamed property, will return:
```json
{
  "additionalData": {},
  "acrValuesSupported": [],
  "authorizationEndpoint": "https://localhost:4000/connect/authorization",
  "checkSessionIframe": null,
  "claimsSupported": [],
  "claimsLocalesSupported": [],
  "claimsParameterSupported": false,
  "claimTypesSupported": [],
  "displayValuesSupported": [],
  "endSessionEndpoint": null,
  "frontchannelLogoutSessionSupported": null,
  "frontchannelLogoutSupported": null,
  "grantTypesSupported": [],
  "httpLogoutSupported": false,
  "idTokenEncryptionAlgValuesSupported": [],
  "idTokenEncryptionEncValuesSupported": [],
  "idTokenSigningAlgValuesSupported": [],
  "introspectionEndpoint": null,
  "issuer": "https://localhost:4000",
  "jwks_uri": "https://localhost:4000/.well-known/openid-configuration/jwks",
  "jsonWebKeySet": null,
  "logoutSessionSupported": false,
  "opPolicyUri": null,
  "opTosUri": null,
  "registrationEndpoint": null,
  "requestObjectEncryptionAlgValuesSupported": [],
  "requestObjectEncryptionEncValuesSupported": [],
  "requestObjectSigningAlgValuesSupported": [],
  "requestParameterSupported": false,
  "requestUriParameterSupported": false,
  "requireRequestUriRegistration": false,
  "responseModesSupported": [],
  "responseTypesSupported": [],
  "serviceDocumentation": null,
  "scopesSupported": [],
  "signingKeys": [],
  "subjectTypesSupported": [],
  "tokenEndpoint": "https://localhost:4000/connect/authentication",
  "tokenEndpointAuthMethodsSupported": [],
  "tokenEndpointAuthSigningAlgValuesSupported": [],
  "uiLocalesSupported": [],
  "userInfoEndpoint": null,
  "userInfoEndpointEncryptionAlgValuesSupported": [],
  "userInfoEndpointEncryptionEncValuesSupported": [],
  "userInfoEndpointSigningAlgValuesSupported": []
}
```

I did not investigate the reason the attribute JsonProperty is not serializing with the specified PropertyName, for now the implementation is running as expected with the workaround.

Let me know if I can give any further information.

**Comment by brentschmaltz** [comment]
@RogerioWagner how are you serializing to Json? I ask because we have roundtrip tests here: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/a22c4f474b806c7c90f6dd9b5fd371fd55db51d4/test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/OpenIdConnectConfigurationTests.cs#L234

**Comment by RogerioWagner** [comment]
@brentschmaltz I'm not serializing by myself, the json was captured in postman calling the endpoint https://localhost:4000/.well-known/openid-configuration, which is routed to the method you can see on my previous post.

**Comment by kashtak** [comment]
I'm stuck with the same issue recently. 
Just want to add, that in the OpenIdConnect library version 5.2.0 the Newtonsoft JsonProperty attribute is used to specify that "jwks_uri" name. And it works (serializes an object) correctly.
Since the OpenIdConnect started (don't know exact version) to use JsonProperty attribute from the System.Text.Json library it ignores the specified property name.
I don't serialize the OpenIdConnectConfiguration object explicitly - just `return Json(configuration);` from my controller.

**Comment by brentschmaltz** [comment]
@RogerioWagner @kashtak with the 5.x releases, you could add the NewtonSoft directly. With the 6.x release, we removed the public exposure to NewtonSoft and you scenario is now broken.

ensure: #1496  #1465 are fixed also.

**Comment by chwarr** [comment]
It appears to me that we can access the "internal" forked Newtonsoft.JSON serializer by calling [`System.IdentityModel.Tokens.Jwt.JsonExtensions.SerializeToJson()`](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/30c16ab82cf4338c6d6e04a5d4094795aa83c587/src/System.IdentityModel.Tokens.Jwt/JsonExtensions.cs#L88-L96).

@RogerioWagner, if you manually invoke that in your `GetConfiguration()` method and then do whatever to need to do to return a "raw" payload, can you get it to work?

**Comment by cleferman** [comment]
Stumbled upon this today and wrecked my head for a few hours until I finally retrieved the discovery document "manually".
I was sure I configured the JWT middleware correctly but it never hit my /keys endpoint. I inspected the url a dozen times, it was correct but it turns out it was a serialization issue...

I'll just leave this chunk of code here
```csharp
        [HttpGet]
        public IActionResult GetConfiguration(CancellationToken cancellationToken)
        {
            var requestedUrlBase = $"{Request.Scheme}://{Request.Host}/";

            var configuration = new OpenIdConnectConfiguration
            {
                Issuer = securitySettings.Issuer,
                TokenEndpoint = requestedUrlBase + "token",
                JwksUri = requestedUrlBase + "keys",
            };

            return Ok(configuration);
        }
```
I looked into the OkObjectResult class and it seems it's using the System.Text.Json serializer.
I tried forcing the controllers to use the Newtonsoft.Json one but with no success `services.AddControllers().AddNewtonsoftJson();`

Inheriting from the OpenIdConnectConfiguration class and adding the jwks_uri property as @RogerioWagner  mentioned works like a charm.

Sorry for bumping this, but it still seems to be an issue. Here's to hoping the next guy finds this before losing their minds like I almost did.
